src/app/clusters/time-format-localization-server/time-format-localization-server.cpp

```diff
@@ -0,0 +1,222 @@
+
+#include <app-common/zap-generated/af-structs.h>
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/AttributeAccessInterface.h>
+#include <app/util/attribute-storage.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <platform/PlatformManager.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::TimeFormatLocalization;
+using namespace chip::app::Clusters::TimeFormatLocalization::Attributes;
+using namespace chip::DeviceLayer;
+
+namespace {
+
+class TimeFormatLocalizationAttrAccess : public AttributeAccessInterface
+{
+public:
+    // Register for the Time Format Localization cluster on all endpoints.
+    TimeFormatLocalizationAttrAccess() : AttributeAccessInterface(Optional<EndpointId>::Missing(), TimeFormatLocalization::Id) {}
+
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+
+private:
+    CHIP_ERROR ReadSupportedCalendarTypes(AttributeValueEncoder & aEncoder);
+};
+
+TimeFormatLocalizationAttrAccess gAttrAccess;
+
+CHIP_ERROR TimeFormatLocalizationAttrAccess::ReadSupportedCalendarTypes(AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    DeviceLayer::AttributeList<TimeFormatLocalization::CalendarType, DeviceLayer::kMaxCalendarTypes> supportedCalendarTypes;
+
+    if (DeviceLayer::PlatformMgr().GetSupportedCalendarTypes(supportedCalendarTypes) == CHIP_NO_ERROR)
+    {
+        err = aEncoder.EncodeList([&supportedCalendarTypes](const auto & encoder) -> CHIP_ERROR {
+            for (auto type : supportedCalendarTypes)
+            {
+                ReturnErrorOnFailure(encoder.Encode(type));
+            }
+
+            return CHIP_NO_ERROR;
+        });
+    }
+    else
+    {
+        err = aEncoder.EncodeEmptyList();
+    }
+
+    return err;
+}
+
+CHIP_ERROR TimeFormatLocalizationAttrAccess::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    VerifyOrDie(aPath.mClusterId == TimeFormatLocalization::Id);
+
+    switch (aPath.mAttributeId)
+    {
+    case TimeFormatLocalization::Attributes::SupportedCalendarTypes::Id:
+        return ReadSupportedCalendarTypes(aEncoder);
+    default:
+        break;
+    }
+    return CHIP_NO_ERROR;
+}
+
+} // anonymous namespace
+
+// =============================================================================
+// Pre-change callbacks for cluster attributes
+// =============================================================================
+
+chip::Protocols::InteractionModel::Status __attribute__((weak))
+emberAfPluginTimeFormatLocalizationOnHourFormatChange(chip::EndpointId EndpointId, uint8_t newFormat)
```

Comment:
This is OK, I guess, but for this specific case it seems like min/max values in the XML would get you automatic checking of this so you wouldn't need this function....